pkg/kubelet/server.go

```diff
@@ -620,80 +628,179 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
 		return
 	}
 
-	streamChan := make(chan httpstream.Stream, 1)
-	upgrader := spdy.NewResponseUpgrader()
-	conn := upgrader.UpgradeResponse(response.ResponseWriter, request.Request, func(stream httpstream.Stream) error {
+	podName := kubecontainer.GetPodFullName(pod)
+
+	ServePortForward(response.ResponseWriter, request.Request, s.host, podName, uid, s.host.StreamingConnectionIdleTimeout())
+}
+
+func portForwardStreamReceived(streams chan httpstream.Stream) func(httpstream.Stream) error {
+	return func(stream httpstream.Stream) error {
 		portString := stream.Headers().Get(api.PortHeader)
 		port, err := strconv.ParseUint(portString, 10, 16)
 		if err != nil {
-			return fmt.Errorf("Unable to parse '%s' as a port: %v", portString, err)
+			return fmt.Errorf("unable to parse %q as a port: %v", portString, err)
 		}
 		if port < 1 {
-			return fmt.Errorf("Port '%d' must be greater than 0", port)
+			return fmt.Errorf("port %d must be greater than 0", port)
 		}
-		streamChan <- stream
+		streams <- stream
 		return nil
-	})
+	}
+}
+
+func getPortForwardStreamConnectionID(stream httpstream.Stream) string {
+	connectionID := stream.Headers().Get(api.ConnectionIDHeader)
+	if len(connectionID) == 0 {
+		glog.V(5).Infof("Stream received without connection ID header")
+		// If we get here, it's because the connection came from an older client
+		// that isn't generating the connectionID header
+		// (https://github.com/kubernetes/kubernetes/blob/843134885e7e0b360eb5441e85b1410a8b1a7a0c/pkg/client/unversioned/portforward/portforward.go#L258-L287)
+		//
+		// This is a best-effort attempt at supporting older clients.
+		//
+		// When there aren't concurrent new forwarded connections, each connection
+		// will have a pair of streams (data, error), and the stream IDs will be
+		// consecutive odd numbers, e.g. 1 and 3 for the first connection. Convert
+		// the stream ID into a pseudo-connectionID by taking the stream type and
+		// using connectionID = stream.Identifier() when the stream type is error,
+		// and connectionID = stream.Identifier() - 2 when it's data.
+		//
+		// NOTE: this only works when there are not concurrent new streams from
+		// multiple forwarded connections; it's a best-effort attempt at supporting
+		// old clients that don't generate connection IDs.  If there are concurrent
+		// new connections, it's possible that 1 connection gets streams whose IDs
+		// are not consecutive (e.g. 5 and 9 instead of 5 and 7).
+
+		streamType := stream.Headers().Get(api.StreamType)
+		switch streamType {
+		case api.StreamTypeError:
+			connectionID = strconv.Itoa(int(stream.Identifier()))
+		case api.StreamTypeData:
+			connectionID = strconv.Itoa(int(stream.Identifier()) - 2)
+		default:
+			connectionID = ""
```

Comment:
Isn't connectionID already `""` here? Also, is this case worth logging or handling as an error?